### ISSUE TITLE
GGML formatted Flash Attention

### ISSUE THREAD

**Issue by prajaramanTT** [tenstorrent/tt-metal#15943]
TTNN provides ttnn::transformer::scaled_dot_product_attention_decode which is very close to
what GGML wants. But with a few major differences
Format of Q is different
GGML wants [1, n_head, n_batch, n_embed]
TTNN wants [1, n_batch, n_head n_embed]
TTNN does not apply ALiBi to the result (though it seems to contain softmax)
Different mask shape
GGML wants [1, 1, n_batch_pad, n_kv]
TTNN wants [1, n_n_batch, s, s]

NOTE: GGML by default disables Flash Attention. Flag has to be set to enable it.
The definition of GGML's modified FlashAttention can be found below. The result GGML tensor is
permuted internally. Which can be ignored in TTNN as it handles views in a different
mechanism.

// q: [n_embd, n_batch, n_head, 1]
// k: [n_embd, n_kv, n_head_kv, 1]
// v: [n_embd, n_kv, n_head_kv, 1] !! not transposed !!
// mask: [n_kv, n_batch_pad, 1, 1] !! n_batch_pad =
GGML_PAD(n_batch, GGML_KQ_MASK_PAD) !!
// res: [n_embd, n_head, n_batch, 1] !! permuted !!
GGML_API struct ggml_tensor * ggml_flash_attn_ext(
struct ggml_context * ctx,
struct ggml_tensor * q,
struct ggml_tensor * k,
struct ggml_tensor * v,
struct ggml_tensor * mask,
float scale,
float max_bias,
float logit_softcap);

To clarify:
Fused Softmax + ALiBi, Generalize MatMul broadcasting

- Not strictly needed if GGML formatted FlashAttention with ALiBi is implemented

- But feels like an incremental step towards the needed FlashAttention

- Also GGML by default disables FlashAttention. So I still think they should be
implemented

The rest are definitely needed for performance.

**Comment by uaydonat** [comment]
@xuncaiTT @cglagovichTT any input how much work it would be to re-implement GGML formatted FlashAttention? Or would that so slow that it does not make sense?

**Comment by cglagovichTT** [comment]
Looks like ALiBi is computed inside the kernel https://github.com/ggerganov/llama.cpp/blob/9c8dcefe171ba70ed14f2a64d1433da12d0dd1c1/ggml/src/ggml-cuda/fattn-tile-f16.cu#L67

**Comment by caixunshiren** [comment]
> cross-referencing [#10376](https://github.com/tenstorrent/tt-metal/issues/10376)
> 
> It's a decent amount of work to make FlashDecode accept the input shapes that GGML expects.
> 
> * We would have to fuse a `transpose(1, 2)` TM in front of Q.
> * Add ALiBi (would be easier if this was passed as an attention mask. is it?)
> 
> It could be easier if we make a composite GGML version of FlashDecode which wraps the existing FlashDecode and applies the TM and generates and ALiBi mask. Thoughts?

I agree. Quite a few changes has to be made to make a GGML version of flash decode, and they are extra steps suboptimal for perf. For functional version, having GGML flash decode as a composite op is probably the best way to do it.

**Comment by marty1885** [comment]
Ok, let me implement it as a composition op first and see how it goes. But IIRC my last experiment shows the transpose is a major pain in performance.

**Comment by ayerofieiev-tt** [comment]
@marty1885 , moved to you. please close if not relevant
